c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+  for (const auto& free_stream : capture_free_streams) {
+    // cudaEventRecord requires that the input event and stream are on the same device.
+    CUDAGuard g(free_stream.device);
+
+    // CUDACachingAllocator.cpp uses raw cuda events, as do we.
+    cudaEvent_t event;
+    C10_CUDA_CHECK(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+    C10_CUDA_CHECK(cudaEventRecord(event, free_stream.stream));
+    C10_CUDA_CHECK(cudaStreamWaitEvent(capture_stream.stream(), event));
+    C10_CUDA_CHECK(cudaEventDestroy(event));
+  }
+
+  capture_free_streams.clear();
+}
+
+void notifyCaptureEnded(int device, CaptureId_t graph_id) {
+  assertValidDevice(device);
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  TORCH_CHECK(capture_underway,
+              "CudaMallocAsync::notifyCaptureEnded called, "
+              "but CudaMallocAsync::capture_underway is false.");
+  capture_underway = false;
+
+  // See Note [Avoid freeing uncaptured ptrs during CUDA graph capture]
+  for (const auto ptr : ungraphed_ptrs_defer_free_until_no_capture) {
+    auto it = ptr_info.find(ptr);
+    TORCH_INTERNAL_ASSERT(it != ptr_info.end(),
+                          "ptr not found in ptr_info");
+    TORCH_INTERNAL_ASSERT(it->second.usage_streams.size() != 0,
+                          "ptr's stream uses vector is empty");
+    free_impl(it);
+  }
+
+  ungraphed_ptrs_defer_free_until_no_capture.clear();
+}
+
+void notifyCaptureDestroy(int device, MempoolId_t mempool_id) {
+  // Q: Do we need to do anything special here, like clear long-lived
+  //    pointers created during the original capture (for example,
+  //    tensors intended as the graph's I/O surface) that might still
+  //    be resident in ptr_info?
+  // A: I don't think so.
+  //    Those allocations survived capture because the user held
+  //    explicit tensor references to them,
+  //    Those tensors' destructors will call free() on each pointer
+  //    when the user is done with them.
+  //    The free()s will probably incur
+  //    TORCH_WARN("Attempting uncaptured free of a captured allocation..."
+  //    but stale ptrs will not permanently leak into ptr_info.
+}
+
+void* raw_alloc(size_t nbytes) {
+  if (nbytes == 0) {
+    return nullptr;
+  }
+  int device;
+  C10_CUDA_CHECK(cudaGetDevice(&device));
+  void* r = nullptr;
+  malloc(&r, device, nbytes, cuda::getCurrentCUDAStream(device));
+  return r;
+}
+
+void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
+  if (nbytes == 0) {
+    return nullptr;
+  }
+  int device;
+  C10_CUDA_CHECK(cudaGetDevice(&device));
+  void* r = nullptr;
+  malloc(&r, device, nbytes, stream);
+  return r;
+}
+
+void raw_delete(void* ptr) {
+  free(ptr);
+}
+
+#else
+// CUDA_VERSION is < 11040
+
+#define NOT_AVAILABLE(name) \
+TORCH_CHECK(false, \
+            "Called CudaMallocAsync::" \
+            name \
+            " but Pytorch was built with cuda < 11.4."); \
```

Comment:
I gave you a pass in the comments, but this is an error message; it's PyTorch not Pytorch lol